File: nelman-kmcLATTICEMODEL/source/GranCan.cpp

#include "GranCan.hpp"
#include "Lattice.hpp"
#include "Chain.hpp"
#include "MolBox.hpp"
#include "EnergyMap.hpp"
#include "Stats.hpp"
#include "StatsMethods.hpp"
#include "Cluster.hpp"
#include "ptemp.h"

#include "App.hpp"





#include <string.h>
#include <math.h>
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <sstream>
#include <fstream>
#include <vector>




extern Pos local[6];

 

GranCan::GranCan(Lattice * ml){

  mainLattice = ml;
  molbox=new MolBox(mainLattice);  // needed to for Stats::setLattice

  Volume = (double) (LX * LY * LZ) ; 
  App::eBetaMu = -0.0000005;
}



int GranCan::trialInsertChain(){
  //cout<< "trial insert chain"<<endl;
  // find free chains
  int l=0;
  for(int c=0;c<mainLattice->nChains;c++){
    if(! mainLattice->chains[c]->hasCext() ){
      l++;
    }
  }  
  int freeChains=l;
 // accept depending on mu


  double prefactor;
  if(LAMBDA_COR){
    prefactor = 1.0/pow(mainLattice->getRealBeta(),3.0/2.0);
  }else{
    prefactor =1.0;
  }
  
  double accRatio =  prefactor*App::eBetaMu * (double) Volume/ ((double)freeChains +1.0);
  if(drand48() < accRatio){

    // get configuration
    Config * config =  molbox->getNewConfig();
    Config * newConfig= new Config;
    newConfig->nTotal = config->nTotal;
    // add random translation to points (could do this in molbox) 
    Pos addP;
    addP.x = (int) floor( LX * drand48());
    addP.y = (int) floor( LY * drand48());
    addP.z = (int) floor( LZ * drand48()); 
    
    for(int n=0;n<config->nTotal;n++){
      //cout<<"config res:"<<n<<endl;
      Pos posN = config->positions[n]+addP;
      posN.periodicBoundary();
      // check if lattice points not occupied and not bound
      if(mainLattice->getResidue(posN) != NULL){
	delete newConfig;
	return -1;
      }
      for(int k=0;k<6;k++){
	Pos posNB = local[k] + posN;
	posNB.periodicBoundary();
	// check if bound, threonine (aa==13) contact does not count
	//if(mainLattice->getResidue(posNB)!= NULL && mainLattice->getResidue(posNB)->aa != 13 ) {
	if(mainLattice->getResidue(posNB)){
	  delete newConfig;
	  return -1;
	}
      }
#ifdef CROSSNBS
      Pos spin = config->spins[n];
      Pos * crossNbs = Stats::getCrossNbsSpinDir(spin);
      for(int k=0;k<4;k++){
	Pos posNB =crossNbs[k]+ posN;
	posNB.periodicBoundary();
	Residue * resNB = mainLattice->getResidue(posNB);
	if(resNB!=NULL){
	  bool sameChain = false;
	  if(!sameChain){
	    bool spinsOpposite= (resNB->spin == (Pos(0,0,0)- spin));
	    if(spinsOpposite){
	      delete newConfig;
	      return -1;
	    }
	  }
	}
      }
      //delete crossNbs;
#endif
      //Note this could be more efficient
      newConfig->positions[n]= posN;
      newConfig->spins[n]=config->spins[n];
      newConfig->states[n]=config->states[n];
    }
    
    
    // insert, by creating new chain
    mainLattice->insertChain(newConfig,molbox->sequence,mainLattice->getBetaMoves());
    delete newConfig;
    // mainLattice->writePDBMultiChain("newlyinsertedchain.pdb");
    return 1;
  }
  return 0;
}

// this may become slow for fully aggregated box
int GranCan::trialDeleteChain(){
  // cout<< "trial delete chain"<<endl;
  int freeChns[MAX_CHAINS];
  // find free chains
  int l=0;
  for(int c=0;c< mainLattice->nChains;c++){
    if(! mainLattice->chains[c]->hasCext()){
      freeChns[l]= c;
      l++;
    }
  }  
  int freeChains=l;
  if(freeChains < 1) return -2;
  if(mainLattice->nChains <=1 ){ 
#ifdef CORRECTION_0CHAINS
    double prefactor;
    if(LAMBDA_COR){
       prefactor = pow(mainLattice->getRealBeta(),3.0/2.0);
    }else{
      prefactor =1.0;
    }

   
    double accRatio = prefactor*((double)freeChains ) / (App::eBetaMu *  Volume);
    if(drand48() < accRatio){
      
      double alpha = App::eBetaMu * ((double) Volume)/prefactor;
      //double r = drand48();
      double num_steps;
      if(alpha > 0.9999){
	num_steps=1;
      }else{
	num_steps = ceil(log(drand48())/log(1.0-alpha));
      }
      double moves = (1.0+ pGlobal)*(2*num_steps/pInsDel);
      totalEmptySteps += moves;
      //cout<<"Volume "<<Volume<<endl;
      //cout<<"r "<<r<<endl;
      //cout<<"alpha "<<alpha<<endl;
      //cout<<"log(1.0-alpha)"<<log(1.0-alpha)<<endl;
      //cout<<"pG "<<pGlobal<<endl;
      //cout<<"n_s "<<num_steps<<endl;
      //cout<<"moves "<<moves<<endl;
      Stats emptyStats ;
      mainLattice->energyMap->mapStats(emptyStats,moves);
      
    }
    return -1;
#else
    return -1;
#endif
  }
  /*do{  
    chain = (int) floor(drand48() * nChains);  
    }while(chains[chain]->Cext!=0);
  */
	
  // accept deletion depending on mu
  double prefactor;
  if(LAMBDA_COR){
    prefactor = pow(mainLattice->getRealBeta(),3.0/2.0);
  }else{
    prefactor =1.0;
  }


  double accRatio = prefactor*((double)freeChains ) / (App::eBetaMu *  Volume);
  if(drand48() < accRatio){
     // pick a free chain at random
   
    int chain= freeChns[(int) floor(drand48() * freeChains)];
    // delete chain from lattice
    mainLattice->deleteChain(chain);
    return 1;
  }
  return 0;
}

int GranCan::getNumFreeChains(){
  // cout<<"getNumfree chains" <<endl;
 int l=0;
  for(int c=0;c< mainLattice->nChains;c++){
    if(! mainLattice->chains[c]->hasCext()){
      l++;
    }
  } 
  return l;
}





int GranCan::clusterMove(){
  /// calculate clusters
  Cluster cl;
  cl.startCounting(mainLattice);
  int numClusts = cl.getTotalClusters();
  //  cl.printCout();

  /// pick random clusters
  int cluster2move = (int) floor(numClusts * drand48());

  /// get all chains in cluster
  vector<int> chainList = cl.getChains(cluster2move);
  int numChains = chainList.size();

  /// set rotation and translation
  //int dir_translation= (int)floor(6 * drand48());
  //int magnitude_translation = (int)floor(MAX_TRANSLATION * drand48());
  int rotationalDir= (int)floor(6 * drand48());

  /// choose random pivot on lattice
  int xpivot =  (int)floor(LX * drand48()); 
  int ypivot =  (int)floor(LY * drand48());
  int zpivot =  (int)floor(LZ * drand48());
  Pos posPivot =  Pos(xpivot,ypivot,zpivot);

  vector< vector <Pos> > newPos (numChains, vector<Pos>());

  /// for each chain, perform rotation and translation,
  /// while keeping old coordnates
  /// check for clashes or touches: reject, restore old coordinates and sample
  for(unsigned int i=0; i< chainList.size();i++){
    int cn = chainList[i];
    Chain * ch = mainLattice->chains[cn];
    for(int n=0; n<ch->N; n++){
      Residue * res = ch->residues[n];
      Pos pp = ch->rotationPosition(res->pos,rotationalDir,posPivot);
      newPos[i].push_back(pp);
      if(mainLattice->checkForClashesAndTouches(newPos[i][n])){
	/// reject move
	// l->sampleCurrentStats();
	return -1;
      }      

    }
  }
  /// now we know there are no clashes our touches
  /// and the move can be performed
  /// note that the move will alsways get accepted, since dE=0

  /// Perform move, update lattice
  for(unsigned int i=0; i< chainList.size();i++){
    int cn = chainList[i];
    Chain * ch = mainLattice->chains[cn];
    for(int n=0; n<ch->N;n++){
      Residue * res = ch->residues[n];
      Pos newSpin = ch->rotationSpin(res->spin,rotationalDir);
      mainLattice->emptyLatticePos(res->pos);

      res->pos = newPos[i][n];
      res->spin = newSpin;
      mainLattice->setResidue(newPos[i][n],res);
    }
    ch->setFwdBkwdSpins();
  }
  return 1;
}

File: nelman-kmcLATTICEMODEL/source/GranCan.py
from typing import List, Optional
import random
import math
from Lattice import Lattice
from MolBox import MolBox
from Cluster import Cluster
from Pos import Pos
from Config import Config
from Residue import Residue

MAX_CHAINS = 50000
MAX_TRANSLATION = 10

class GranCan:
    def __init__(self, main_lattice: Lattice):
        self.main_lattice = main_lattice
        self.molbox = MolBox(main_lattice)
        self.volume = main_lattice.LX * main_lattice.LY * main_lattice.LZ
        self.p_global = 0.0
        self.p_ins_del = 0.0
        self.p_change_strand_coil = 0.0
        self.p_cluster_move = 0.0
        self.steps = 0
        self.total_swaps = 0
        self.total_empty_steps = 0.0
    
    def trialInsertChain(self) -> int:
        # Find free chains
        free_chains = [c for c in range(self.main_lattice.nChains) 
                      if not self.main_lattice.chains[c].hasCext()]
        num_free = len(free_chains)
        
        # Calculate acceptance ratio
        prefactor = 1.0 / math.pow(self.main_lattice.getRealBeta(), 1.5) if App.LAMBDA_COR else 1.0
        acc_ratio = prefactor * App.eBetaMu * self.volume / (num_free + 1.0)
        
        if random.random() < acc_ratio:
            config = self.molbox.getNewConfig()
            if config is None:
                return -1
                
            # Add random translation
            add_p = Pos(
                random.randint(0, self.main_lattice.LX - 1),
                random.randint(0, self.main_lattice.LY - 1),
                random.randint(0, self.main_lattice.LZ - 1)
            )
            
            # Check for clashes
            for n in range(config.nTotal):
                pos_n = config.positions[n] + add_p
                pos_n.periodicBoundary()
                
                if self.main_lattice.getResidue(pos_n) is not None:
                    return -1
                    
                for k in range(6):
                    pos_nb = Pos.local[k] + pos_n
                    pos_nb.periodicBoundary()
                    if self.main_lattice.getResidue(pos_nb) is not None:
                        return -1
            
            # Insert the chain
            self.main_lattice.insertChain(config, self.molbox.sequence, self.main_lattice.getBetaMoves())
            return 1
            
        return 0
    
    def trialDeleteChain(self) -> int:
        free_chains = [c for c in range(self.main_lattice.nChains) 
                      if not self.main_lattice.chains[c].hasCext()]
        num_free = len(free_chains)
        
        if num_free < 1:
            return -2
        if self.main_lattice.nChains <= 1:
            return -1
            
        # Calculate acceptance ratio
        prefactor = math.pow(self.main_lattice.getRealBeta(), 1.5) if App.LAMBDA_COR else 1.0
        acc_ratio = prefactor * num_free / (App.eBetaMu * self.volume)
        
        if random.random() < acc_ratio:
            chain = random.choice(free_chains)
            self.main_lattice.deleteChain(chain)
            return 1
            
        return 0
    
    def getNumFreeChains(self) -> int:
        return sum(1 for c in range(self.main_lattice.nChains) 
                 if not self.main_lattice.chains[c].hasCext())
    
    def clusterMove(self) -> int:
        # Similar to cluster_move function above
        cl = Cluster()
        cl.startCounting(self.main_lattice)
        num_clusters = cl.getTotalClusters()
        
        if num_clusters == 0:
            return -1
            
        cluster_to_move = random.randint(0, num_clusters - 1)
        chain_list = cl.getChains(cluster_to_move)
        
        rotational_dir = random.randint(0, 5)
        pivot = Pos(
            random.randint(0, self.main_lattice.LX - 1),
            random.randint(0, self.main_lattice.LY - 1),
            random.randint(0, self.main_lattice.LZ - 1)
        )
        
        # Check for clashes
        new_positions = []
        for cn in chain_list:
            ch = self.main_lattice.chains[cn]
            chain_pos = []
            for n in range(ch.N):
                res = ch.residues[n]
                new_pos = ch.rotationPosition(res.pos, rotational_dir, pivot)
                chain_pos.append(new_pos)
                if self.main_lattice.checkForClashesAndTouches(new_pos):
                    return -1
            new_positions.append(chain_pos)
        
        # Perform the move
        for i, cn in enumerate(chain_list):
            ch = self.main_lattice.chains[cn]
            for n in range(ch.N):
                res = ch.residues[n]
                new_spin = ch.rotationSpin(res.spin, rotational_dir)
                self.main_lattice.emptyLatticePos(res.pos)
                res.pos = new_positions[i][n]
                res.spin = new_spin
                self.main_lattice.setResidue(new_positions[i][n], res)
            ch.setFwdBkwdSpins()
        
        return 1
